log: adsp_hda: Fixes immediate mode test

teburd opened the pull request:
Make the hda dma logger a pipeline with concurrent format outputs. Avoids IPC usage when used with cavstool, the only requirement is HDA, logging, and kernel primitives. Easily the core HDA output streaming functions could probably be broken off to be a nice little reusable output stream if wanted, of any byte data.

Fixes #50404 and #50455 for me and the output even looks decent with many threads.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

shuai1x commented:
I tested it on CAVS25. And this PR dramatically reduces the probability of failure. But It is still having the problem that I raised. I tried five times and got one failure, but this doesn't mean the odds are one in five. Thanks.

teburd commented:
There's another problem I've ran into here I don't have an explanation for yet

The host marking an IPC as "DONE" never seems to be getting to the DSP in some cases.

teburd commented:
A new version of this that changes the behavior quite a bit. There's some printk/hardcoded cavstool log level set so when people are testing it out if there's still problems hopefully we can figure out why faster. Once I hear it fixes the problem @shuai1x I'll clean all that up.

Please try this updated patch and see if issues are solved.

teburd commented:
The latest changeset tries maybe a few too many things to resolve the issue but...

I can run tests many many times without fail, though they are quite slow due to the sleep in replying to the IPC in cavstool. Without this sleep, of this length, I eventually fail to receive an IPC reply marking the DONE bit on the DSP side.

teburd commented:
Ok latest commit is pretty close I think, there's still some race somewhere that seems to just cause the whole thing to become a smoking wreck of bits.

teburd commented:
Some observations worth noting...

**Working Configs**
CONFIG_MP_NUM_CPUS = 4, NUM_THREADS = 3
CONFIG_MP_NUM_CPUS = 1, NUM_THREADS = 5

**Failing Configs**
CONFIG_MP_NUM_CPUS = 4, NUM_THREADS = 5
Uncommenting the flush timer in adsp_hda

An unlucky interruption seems to be a potential culprit, once it happens the train wreck.

shuai1x commented:
> A new version of this that changes the behavior quite a bit. There's some printk/hardcoded cavstool log level set so when people are testing it out if there's still problems hopefully we can figure out why faster. Once I hear it fixes the problem @shuai1x I'll clean all that up.
> 
> Please try this updated patch and see if issues are solved.

Ok, I will test it again and reply to you soon.

shuai1x commented:
> A new version of this that changes the behavior quite a bit. There's some printk/hardcoded cavstool log level set so when people are testing it out if there's still problems hopefully we can figure out why faster. Once I hear it fixes the problem @shuai1x I'll clean all that up.
> 
> Please try this updated patch and see if issues are solved.

I have run ten times on CAVS25, and all can pass. The issues are solved. Thanks.

teburd commented:
Well, sort of. In fixing this though I've uncovered a more troublesome issue I think, but at least the patch works and solves the issues so we know what does work now.

Commit:
soc: intel_adsp: cavstool: Update SPIB for HDA Streams

Update the software position in buffer for the HDA Stream each
time a log message is read and printed from it.

Configures SPIB to be used (SPIBE) each time a stream is configured.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
logging: adsp_hda: Pipelined, concurrent, correct

Treats the entire thing like a little pipe line of tasks from copying
the log to the dma buffer, padding, flushing, and notifying.

Each step has a pending byte count. The available byte count is always
known. The current write position is always known. Probing the DMA
status is only done to verify writes. Each step may be concurrently
happening across processors.

Log formatting is done on a small pool of outputters for threads and by
a single output protected by a lock for ISRs, panic flushes, and dropped
messages in immediate mode.

Sets a minimum bound on flush and hooks to avoid a fault
where, if using IPC especially, the IPC registers will fail to
update.

Cavstool now has a few seperate async poll loops which seems to make a
modest difference though a real thread without the GIL would probably be
ideal. It also includes a hda log polling task to avoid using IPCs
entirely for the HDA logger.

The hda log test case has been updated to concurrently write from ISRs
and thread in immediate, deferred, printk, and single cpu
configurations. It also prints some nice information about throughput.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
I think that's all the gremlins finally...

The test case now really hammers this logger in immediate mode from ISRs, threads, and so on.

kv2019i commented on subsys/logging/log_backend_adsp_hda.c:
```diff
@@ -245,17 +333,81 @@ static int format_set(const struct log_backend *const backend, uint32_t log_type
+		}
+	} while (local == -1);
+
+	log_output_func(&outputs[local], &msg->log, flags);
```
Ok this is quite clever and probably something we need to copy to other backends as well. I wonder if we could share this code somehow so it is easier to reuse. It seems this is needed by any backend that runs on multicore systems to avoid blocking interrupts for long periods of time.

teburd replied:
It did require a bit more finesse than this it turns out, but it works really really well with LOG_MODE_IMMEDIATE. In effect every core independently deals with formatting now and the throughput of the logger went from something like 8-20k/s to ~500k/s with 8 threads on cavs25.

kv2019i reviewed (changes requested):
Looks very promising! The optimization on log_output seems like something we should use across the board in backends (at least as an opt-in feature). 

I did find one bug with use of uptime_delta. Please check.

teburd commented:
I gave up on IPC, touching it was endless problematic. Instead poll the host side FIFO position register periodically and get logs without fail every time.

teburd commented:
not really needed anymore with #50623

Might be worth getting CI to pass and keeping though to work on integrating some of the ideas into other smp friendly log backends.

teburd commented:
@nordic-krch is this tactic of having a pool of output buffers to format into something that could be moved up a level into the log subsys so the backends don't need to deal with this?
